### PR TITLE
chore: Docker refactor 🐳

### PR DESCRIPTION
**Pull request by neilsong**

### CODE REVIEW

**Comment by neilsong** [comment]
@bearsyankees I need access to either production or a clone of production's VM to test the production docker setup. After I confirm that to be working, I will delete `docker/`, add  update docs, and we can ship.

Sidenote: Infra is very poorly designed. Each folder should not have a separate compose project, this defeats the entire purpose of Docker Compose. Instead, each service should be referenced in a root docker-compose.yml, with service-specific details defined in their own Dockerfiles. (See: my modifications to `legacy/` and `mysql/` in `api/`)

In production, we maintain microarchitecture isolation by modifying the `build` and `up` commands to target individual services when rebuilding/redeploying for max uptime. 

If you think it's appropriate, I can redesign infra, test on that separate VM, and open PR.

**Comment by Josh-Cena** [comment]
@neilsong I tried resolving the conflicts. Please check the diff: https://github.com/coursetable/coursetable/pull/1351/commits/9c773f1dea1ce93b38854ff03eaafc332b101145 to see if I did it right

**Comment by neilsong** [comment]
> @neilsong I tried resolving the conflicts. Please check the diff: [9c773f1](https://github.com/coursetable/coursetable/commit/9c773f1dea1ce93b38854ff03eaafc332b101145) to see if I did it right

LGTM except the Dockerfile - pushed modified changes.

**Comment by neilsong** [comment]
> @bearsyankees I need access to either production or a clone of production's VM to test the production docker setup. After I confirm that to be working, I will delete `docker/`, add update docs, and we can ship.
> 
> Sidenote: Infra is very poorly designed. Each folder should not have a separate compose project, this defeats the entire purpose of Docker Compose. Instead, each service should be referenced in a root docker-compose.yml, with service-specific details defined in their own Dockerfiles. (See: my modifications to `legacy/` and `mysql/` in `api/`)
> 
> In production, we maintain microarchitecture isolation by modifying the `build` and `up` commands to target individual services when rebuilding/redeploying for max uptime.
> 
> If you think it's appropriate, I can redesign infra, test on that separate VM, and open PR.

@bearsyankees Although if we're still manually deploying prod, we can merge this as a dev Docker rework and open a new PR for prod. This PR maintains backwards compat for current prod.

**Comment by Josh-Cena** [comment]
Why does legacy even need to be touched? I thought it's just there for historical reasons.

**Comment by neilsong** [comment]
No clue but I'd rather integrate it properly with the Docker Compose and keep things how they are. If we can delete it, I'd rather do that tbh.

File: api/Dockerfile
Comment (bearsyankees) [inline api/Dockerfile:1]: may need to add one more eaccess thing i saw earlier here but this seems better

File: api/docker-compose.yml
Comment (bearsyankees) [inline api/docker-compose.yml:1]: can talk about removing legacy tomorrow

**Review by bearsyankees — COMMENTED** [review]
yea will have to test tomorrow but it looks like great work

File: api/docker-compose.yml
Comment (Josh-Cena) [reply api/docker-compose.yml:1]: I think for historical purposes (especially the blog), it shouldn't be removed, but moved to a separate repo and get archived.

**Pull request merged**